Support registering arbitrary number of contributors to observers

nilsdeppe opened the pull request:
## Proposed changes

- Print useful error message if no test case found in file. I've found an edge case where you basically get no error message that's of any use when creating a new test case. I'm not sure how many people have encountered it, but I have a few times now and each time I'm confused by it.
- Support registration of multiple different observers. Basically, add strings as a way to identify `ObservationId`s and an `ObservationIdKey` class that is a key into an unordered_map. Events can specify a function that returns an `ObservationId` and `TypeOfObservation` that is then used to register with the observers.

I ended up having to comment out for of the elliptic tests because I have no clue how the elliptic code does observations. It's very different from the evolution code and is not at all easy to follow (missing includes + lots of type aliases).

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Support registration of multiple different observers

nilsdeppe committed:
Print useful error message if no test case found in file

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/ConjugateGradient/CMakeLists.txt:
```diff
@@ -20,5 +20,5 @@ add_dependencies(
 
-add_linear_solver_algorithm_test("ConjugateGradientAlgorithm")
-add_linear_solver_algorithm_test("DistributedConjugateGradientAlgorithm")
+# add_linear_solver_algorithm_test("ConjugateGradientAlgorithm")
```
It might be kind to at least add a comment explaining; maybe something like `# temporarily commented out until the observer registration changes are appropriately propagated to the linear solver code`

moxcodes commented on src/IO/Observer/Actions/RegisterWithObservers.hpp:
```diff
@@ -0,0 +1,70 @@
+ * `std::pair<observers::TypeOfObservation, observers::ObservationId>`
+ */
+template <typename RegisterHelper>
+struct RegisterWithObservers {
```
I think the documentation should be made a little more verbose. It's unclear whether the additional details should go here in the dox for the main public interface, or somewhere in the dev guide, and it's probably better to handle this in a follow-up or just make an issue (to avoid making this PR any longer...). However I think it would be nice to document what the registration with observers means and is doing; and for what uses it is required.

nilsdeppe replied:
I agree. I think we need a dev guide entry giving a complete overview of the observers. There are some more minor refactors and moving files around that would make the code a lot easier to work with. A related thing that I would like to do is to use the observer group and nodegroup as basically the primary groups that we use for everything, not just observation. I'll see if it makes sense to already start the dev guide entry with this PR

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const observers::ObservationId& observation_id,
+                    ArrayComponentId id_of_caller) noexcept {
```
Should this be a const reference, or was intended to be `std::move`d into the next calls?

nilsdeppe replied:
Currently trivially copyable, making `const&` to avoid having to `NOLINT` a bunch of things

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+
+#include <cstddef>
+#include <unordered_map>
+
```
`#include <set>`
`#include <type_traits>`
`#include <unordered_set>`

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
```
I think the macro from `CreateHasTypeAlias.hpp` is only used in one of these action files.

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+                                   std::unordered_set<ArrayComponentId>>*>
+                array_component_ids) noexcept {
+          const ObservationIdRegistrationKey key{observation_id};
+          first_call_for_observation_id =
```
Could the logic here be a little simpler if the boolean was reversed, storing instead `observation_id_already_registered` or similar? I'm hoping to avoid the unfortunate `UNLIKELY(not(...`, instead having something like:
`UNLIKELY(observation_id_already_registered and array_component_ids->at(key).find(component_id) != array_component_ids->at(key).end())` (or the corresponding `.count(component_id) != 0` if that's preferable for concision.

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+      return;
+    }
+
+    const auto register_reduction = [&array_index, &cache,
```
previously, these made sense as separate lambdas because they were called more than once in the switch below. Now, however, they are each only called once, so would it make more sense to just put the simple action calls inline in the switch statement (and maybe get the `observer_writer` before the switch)?

kidder replied:
I agree with this, in fact the only difference is the template parameter of `Parallel::simple_action`

moxcodes commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,98 @@
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
+
+namespace observers {
```
is this one of the cases where we're supposed to use something like `namespace observers::Actions {`?

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+              Parallel::simple_action<
+                  Actions::RegisterReductionContributorWithObserverWriter>(
+                  Parallel::get_parallel_component<
+                      ObserverWriter<Metavariables>>(cache)[0],
```
Now that the node id version is a totally different `apply` function, would it work to just make it a different action altogether? I think that would make the control-flow clearer than using overloads if it doesn't break anything elsewhere.

kidder replied:
I agree with this

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+struct RegisterSenderWithSelf {
+  template <typename ParallelComponent, typename DbTagList,
+            typename Metavariables, typename ArrayIndex,
+            Requires<tmpl::list_contains_v<
```
can this `Requires` become an `if constexpr` too?

moxcodes commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,98 @@
+          using EventType = typename decltype(event_type_v)::type;
+          if constexpr (detail::has_observation_registration_tags_v<
+                            EventType>) {
+            const auto* const derived_class_ptr =
```
If I understand correctly, this block of code with some type magic appears to be extracting the typelist `observation_registration_tags` associated with a particular run-time type information (i.e. the corresponding event that appears in the `trigger_and_events`).  This seems like somewhat scary design, and I'm having trouble finding anywhere in this PR that the full generality covered here is needed (all of the `get_observation_type_and_id_for_registration` functions seem to take 0 arguments in this PR). In particular:
- Could you explain why you anticipate we need this generality down the road?
- Does this break if we have nontrivial inheritance chains in our events?
- Can this be handled with less generality but also less scary trickery by making the `get_observation_type_and_id_for_registration` a free (possibly friend?) template function that takes the event as an argument as well as information from the box in order to create the observation information.

alternatively, if I'm totally missing the point, would you mind explaining what I've misunderstood about this snippet?

nilsdeppe replied:
Clearly I need to improve the documentation. Here's what's happening:

Events can do anything, not just observations, so we need a way of getting the list of/looping over events that will be doing observations. We do this first getting the list of all events that are allowed. Then we loop over all triggers, for each trigger, loop over all the events. We then loop over the possible events and, if that event has a member function that specifies the observation ID needed for registration, try to downcast the current base class event to that derived class. If that works, we perform the registration of the event, if not, then we don't.

Some things that I think are in general necessary to support in registration:
- any number of events
- the observation IDs may depend on something in the DataBox
- this has nothing to do with how complicated the inheritance change is and everything to do with whether the final class in the chain (or the one being used in the factory to create events) has the member function `get_observation_type_and_id_for_registration`. So nontrivial inheritance is fine
- I don't understand what you are thinking with a free function. I don't see how that would work since the observation ID depends on both DataBox info and runtime arguments passed in from the input file.

moxcodes replied:
Ah, okay, that explanation helps a great deal.

Let me see if I can make more concrete my inheritance concern, though perhaps it is a case that should not come up in practice -- at any rate, if our design prohibits it, we should document the constraint.
Consider the scenario in which we have two events `BaseEvent` and `DerivedEvent`, such that `BaseEvent` defines a `get_observation_type_and_id_for_registration`, and `DerivedEvent` inherits the function, and both `BaseEvent` and `DerivedEvent` are in the `possible_events`, but the input file specifies only a single `DerivedEvent` member to be triggered. Now my understanding of this code is (and I'm happy to be corrected):
- the downcast will succeed both when `EventType` is `DerivedEvent` and when it is `BaseEvent` (and both will appear in the loop over `possible_events`
- the registration will occur twice, calling the `BaseEvent` version of the member `get_observation_type_and_id_for_registration` once, and the `DerivedEvent` version once (unless it's a virtual function...).
Am I mistaken?

nilsdeppe replied:
Sure, I view this as unsupported behavior. It would be reasonably easy to check that only one downcast is possible per class being registered.

moxcodes replied:
That sounds good -- if you could add an error in the case that multiple downcasts succeed, I think that would satisfy my concerns.

moxcodes commented on src/IO/Observer/Actions/RegisterSingleton.hpp:
```diff
@@ -0,0 +1,97 @@
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
```
I'll just say 'check includes' here, I think there are probably some missing/some uneeded in this file as well.

moxcodes commented on src/IO/Observer/Initialize.hpp:
```diff
@@ -40,45 +40,30 @@ struct Initialize {
+    }
   }
 
- private:
```
why isn't the helper private anymore?

moxcodes commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
-                    const ArrayIndex& /*array_index*/,
+                    const ArrayIndex& array_index,
                     const observers::ObservationId& observation_id,
+                    ArrayComponentId sender_array_id,
```
Unsure if you want the `ArrayComponentId`s to be const refs, but one way or another it should be consistent (same goes for below)

moxcodes commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
       Requires<
```
can the requires here be `if constexpr`s? (below also)

moxcodes commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
+ * \brief Gathers all the reduction data from all processing elements/cores on a
+ * node.
+ */
+struct CollectReductionDataOnNode {
```
I like this change of reducing the data on the node before sending it to node 0!
I'm curious though -- how much use is it actually to have it be a threaded action when almost all of the work it does is between the lock and unlock? Or is it just that we want to follow the philosophy that if it is on a nodegroup it should be using threaded actions regardless of how much of the body of the action needs to be locked?

nilsdeppe replied:
If memory serves correctly this was being done before. As for the lock, I'd like to replace this with either a simple lock-per-tag or a lock-free hash table (the latter being a rather complicated data structures to use correctly). It's already possible to write and contribute at the same time. I'm going to look into this a bit now. It would be reasonably easy to just change the tag type to a `pair<map, lock>` (especially with the NodeLock wrapper you wrote!). Ultimately, I think it only needs to be possible to write, contribute volume data, and contribute reduction data in parallel. Does that seem reasonable?

nilsdeppe replied:
Ah, the issue is that currently we are only allowed to call `db::mutate` in serial. While the underlying data can be mutated in parallel with various other DataBox operations (the DataBox guarantees pointer stability), we block all calls into the DataBox via a `bool` and a hard error in order to guarantee a persistent valid state. Basically, we can lock the node, retrieve the pointers, unlock the node, lock the data to be manipulated, manipulate the data, unlock the data to be manipulated. It's unclear to me what the cost of locking/unlocking is compared with just waiting for the operations.

moxcodes replied:
Ah, understood. I think given that the possible gains are not easy to reason about, I'm happy with keeping the current implementation if only because it is simpler to read and maintain without the extra locking operations.

nilsdeppe replied:
Well I already had it implemented by the time you added this comment :shrug: now whether there is any performance improvement, :shrug:

moxcodes replied:
heh, too fast for me -- Well, I'm certainly happy with the more granular locking procedure now that you've done all the labor of putting it in :) .

nilsdeppe replied:
:+1: It was surprisingly not too bad to do. Basically two new tags and some non-owning pointers in placing :shrug:

I'm going to write a dev guide for the observer code tomorrow and then push the fixups. There'll still be some maintenance work (@kidder's renames and moving files and actions around), but I'm going to defer that

moxcodes commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
 namespace Tags {
-/// The number of events registered with the observer.
-struct NumberOfEvents : db::SimpleTag {
+/// The observations with an `ObservationIdRegistrationKey` registered
```
Can you expand this comment? It's a bit terse for the reasonably intricate data structure that is represented by this tag.

kidder replied:
I would say
```
/// The set of `ArrayComponentId` registered with each `ObservationIdRegistrationKey`
```

moxcodes commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
+                                  std::unordered_set<ArrayComponentId>>;
+};
+
+/// Used to keep track how many elements have contributed to the reduction, and
```
`how many` -> `which` (unless I misunderstand)

kidder replied:
I would first say 
```
/// \brief The set of `ArrayComponentId` that have contributed to each `ObservationId`
/// for reductions
```
and then have the rest as details

moxcodes commented on src/ParallelAlgorithms/Events/ObserveFields.hpp:
```diff
@@ -133,10 +141,14 @@ class ObserveFields<VolumeDim, ObservationValueTag, tmpl::list<Tensors...>,
                              VariablesToObserve::default_value(),
                          const OptionContext& context = {})
-      : variables_to_observe_(variables_to_observe.begin(),
+      : subfile_path_("/" + subfile_name),
```
Does this break if the "/" is included in the input file option? Maybe clarify in the help text?

moxcodes commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -155,15 +167,39 @@ class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
+  std::pair<observers::TypeOfObservation, observers::ObservationId>
+  get_observation_type_and_id_for_registration() const noexcept {
+    return {observers::TypeOfObservation::Reduction,
+            observers::ObservationId(0.0, subfile_path_ + ".dat")};
```
same question -- should the help text specify  that the `/` and `.dat` should not be included?

moxcodes commented on tests/Unit/Helpers/IO/Observers/ObservationEvent.hpp:
```diff
@@ -0,0 +1,51 @@
+    const db::DataBox<DbTagsList>& box) noexcept {
+  std::pair<::observers::TypeOfObservation, ::observers::ObservationId>
+      result{};
+  tmpl::for_each<typename Event<EventRegistrars>::creatable_classes>(
```
This appears to be functionally the same as the above snippet that I had concerns about in `RegisterEvents.hpp` -- consider moving it to somewhere in `src` and using it as a utility action in `RegistarEventsWithObservers`

moxcodes commented on tests/Unit/IO/Observers/Test_RegisterEvents.cpp:
```diff
@@ -0,0 +1,190 @@
+#include <cstddef>
+#include <string>
+#include <type_traits>
+#include <vector>
```
`<utility>` for `std::pair`, `std::move`, couldn't find where `<vector>` was used in this test.

moxcodes commented on tests/Unit/IO/Observers/Test_RegisterSingleton.cpp:
```diff
@@ -0,0 +1,118 @@
+#include <cstddef>
+#include <string>
+#include <type_traits>
+#include <vector>
```
`<utility>` for `std::pair`, couldn't find where `<string>`, `<vector>`, or `<type_traits>` were used in this test.

moxcodes reviewed (commented):
Several small nitpicks and a few larger design questions.
Cheers!

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
-/// reduction observations.
-struct ReductionArrayComponentIds : db::SimpleTag {
-  using type = std::unordered_set<ArrayComponentId>;
+/// Used to keep track of which nodes have contributed to the particular
```
```
\brief The set of nodes that have contributed to each `ObservationId` for writing reduction data
```
then the rest as details

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
-/// volume observations.
-struct VolumeArrayComponentIds : db::SimpleTag {
-  using type = std::unordered_set<ArrayComponentId>;
+/// The number of ObserverWriter nodegroups that have registered.
```
```
\brief The set of nodes that are registered with each `ObservationIdRegistrationKey` for writing reduction data
```
then the details...

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
+      std::unordered_map<ObservationIdRegistrationKey, std::set<size_t>>;
+};
+
+/// Used to keep track how many elements have contributed to the volume
```
```
/// \brief The set of `ArrayComponentId` that have contributed to each `ObservationId`
/// for volume observation
```
then details...

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -58,6 +91,10 @@ struct ReductionDataNames;
 
 /// Reduction data to be written to disk.
+///
+/// If we have `M` elements we have `M` identical `ObservationId`s sent to the
```
```
/// If we have `M` `ArrayComponentIds` registered with a given `Observer`
///  for a given `ObservationId`, then we expect the `Observer` to receive
/// `M` contributions with the given `ObservationId`.  We combine ...
```

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -73,46 +110,6 @@ struct ReductionDataNames : db::SimpleTag {
-struct ReductionObserversContributed : db::SimpleTag {
-  using type = std::unordered_map<observers::ObservationId, size_t>;
-};
-
```
something I noticed when looking a the documentation, two option tags below are put `\ingroup ObserversGroup` as well as the `OptionTags` namespace.  I suggest removing them from the group

kidder commented on src/IO/Observer/ObservationId.hpp:
```diff
@@ -56,9 +57,18 @@ class ObservationId {
-  template <typename ObservationType>
-  explicit ObservationId(double t, const ObservationType& meta,
-                         const std::string& observation_subtype = "") noexcept;
+  template <typename ObservationType,
```
I would appreciate the following clarifications to the documentation of the class:
```
/// \brief A unique identifier for an observation representing the type of observation and
///  the instance (e.g. time) at which the observation occurs.
///
```
instead of the current brief.  I think the details are sufficient for talking about the H5 file.
Instead of describing examples in words, I think giving code snippets would help.
I would remove the part about the constructor (as they are documented) or move it to the currently undocumented `value` member function

nilsdeppe replied:
- I disagree, I think it's better to document the `time` part in the overall documentation.
- I don't know how to change the words into code.

kidder replied:
- I don't know what you are disagreeing with...
- what words into code?

nilsdeppe replied:
- Where `value`/`t` is documented
- The examples you wanted as code snippets

kidder commented on src/IO/Observer/ObservationId.hpp:
```diff
@@ -99,6 +110,31 @@ bool operator==(const ObservationId& lhs, const ObservationId& rhs) noexcept;
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept;
+
+  size_t key() const noexcept { return key_; }
```
why call this key instead of hash as in ObservationId?

nilsdeppe replied:
To make it clear it's the key to the map, not some hash. That is, this class doesn't do any hashing

kidder commented on src/IO/Observer/ObservationId.hpp:
```diff
@@ -99,6 +110,31 @@ bool operator==(const ObservationId& lhs, const ObservationId& rhs) noexcept;
 
 std::ostream& operator<<(std::ostream& os, const ObservationId& t) noexcept;
+
+/// Used as a key in maps to keep track of how many elements have registered.
```
I would call this `ObservationTypeId` as it is the observation type part of the `ObservationId`
```
/// \brief A unique identifier for representing the type of observation.
///
/// Used for registering observations.

nilsdeppe replied:
This is incorrect. It has nothing to do with the actual type of the observation (volume vs. reduction).

kidder replied:
This is the observation_type_hash part of ObservationId, I know it is not the TypeOfObservation (volume vs reduction) so replace "type of observation" with whatever you think is clear that represents "I am going to be observing whatever the observation_type_hash represents at multiples instances of whatever value represents"

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
+/// The tag is used both on the `Observer` and on the `ObserverWriter`
+/// components since all we need to do is keep track of array component IDs in
+/// both cases.
+struct VolumesContributed : db::SimpleTag {
```
VolumesContributed -> ContributorsOfTensorData ?

nilsdeppe replied:
Feel free to do the renames later if you want. I'm not opposed to the names, just don't have time to do it.

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
+/// We need to keep track of this separately from the local reductions on the
+/// node that are contributing so we need a separate tag. Since nodes are easily
+/// indexed by an unsigned integer, we use `size_t` to keep track of them.
+struct ReductionObserversRegisteredNodes : db::SimpleTag {
```
ReductionObserversRegisteredNodes -> NodesExpectedToContributeReductions ?

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
+/// The tag is used both on the `Observer` and on the `ObserverWriter`
+/// components since all we need to do is keep track of array component IDs in
+/// both cases.
+struct ReductionsContributed : db::SimpleTag {
```
ReductionsContributed -> ContributorsOfReductionData ?

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
-/// The number of events registered with the observer.
-struct NumberOfEvents : db::SimpleTag {
+/// The observations with an `ObservationIdRegistrationKey` registered
+struct ObservationsRegistered : db::SimpleTag {
```
ObservationsRegistered -> ExpectedContributorsForObservations ?

kidder commented on src/IO/Observer/Tags.hpp:
```diff
@@ -26,22 +26,55 @@ namespace observers {
+/// This is used on node 0 (or whichever node has been designated as the one to
+/// write the reduction files). The `unordered_set` is the node IDs that have
+/// contributed so far.
+struct NodeReductionsContributedForWriting : db::SimpleTag {
```
NodeReductionsContributedForWriting -> NodesThatHaveContributedReductions

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+  static void apply(db::DataBox<DbTagsList>& box,
+                    const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const observers::ObservationId& observation_id,
```
why take this and not a ObservationIdRegistrationKey?

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+ * \brief %Actions used by the observer parallel component
+ */
+namespace Actions {
+/// \brief Register a class that will call
```
a class -> an `ArrayComponentId`

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+ */
+namespace Actions {
+/// \brief Register a class that will call
+/// `observers::ThreadedActions::ContributeVolumeData`.
```
at the end add `for the given observers::ObservationId`

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+/// \brief Register a class that will call
+/// `observers::ThreadedActions::ContributeVolumeData`.
+///
+/// Should be invoked on ObserverWriter.
```
at the end add `by the contributor`

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+  }
+};
+
+/// \brief Register a class that will call
```
a class -> an ArrayComponentId

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+
+/// \brief Register a class that will call
+/// `observers::ThreadedActions::WriteReductionData` or
+/// `observers::ThreadedActions::ContributeReductionData`.
```
for the given ObservationId

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+/// `observers::ThreadedActions::WriteReductionData` or
+/// `observers::ThreadedActions::ContributeReductionData`.
+///
+/// Should be invoked on ObserverWriter.
```
by the contributor

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const observers::ObservationId& observation_id,
+                    ArrayComponentId id_of_caller) noexcept {
```
why id_of_caller instead of id_of_contributor as used in the previous action?  Actually I think I prefer id_of_caller in the above action too

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+    if constexpr (tmpl::list_contains_v<
+                      DbTagsList, Tags::ReductionObserversRegisteredNodes>) {
+      const auto node_id = static_cast<size_t>(Parallel::my_node());
+      // processing_element_or_node is the node_id of the caller.
```
I'm not sure what this comment refers to...

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+            volume_observers_registered->at(key).insert(id_of_contributor);
+          });
+    } else {
+      (void)observation_id;
```
(void)box;

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+  static void apply(db::DataBox<DbTagsList>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const observers::ObservationId& observation_id,
```
again, why take this and not a ObservationIdRegistrationKey?

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+            } else {
+              ERROR("Trying to insert a Observer component more than once: "
+                    << id_of_caller
+                    << " with observation id: " << observation_id);
```
why not include this in the error message for the other actions?

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+ * \brief Action that is called on the Observer parallel component to register
+ * the parallel component that will send the data to the observer
+ */
+struct RegisterSenderWithSelf {
```
RegisterSenderWithSelf -> RegisterContributorWithObserver

kidder replied:
I still don't like this name, it makes me wonder why something is registering with itself; but I'm not going to hold anything up...

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+  static void apply(db::DataBox<DbTagList>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& array_index,
+                    const observers::ObservationId& observation_id,
```
again, why take this and not a ObservationIdRegistrationKey?

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+                      array_component_ids->at(key).count(component_id) == 0))) {
+            ERROR(
+                "Trying to insert a component_id more than once for "
+                "observation. This means an element is registering itself "
```
print component_id and observation_id in error message

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+/*!
+ * \brief Action that is called on the Observer parallel component to register
+ * the parallel component that will send the data to the observer
+ */
```
add `Should be invoked on the Observer by the contributor`

kidder commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,270 @@
+
+/*!
+ * \brief Action that is called on the Observer parallel component to register
+ * the parallel component that will send the data to the observer
```
`Register the ArrayComponentId that will send the data to the observer for the given ObservationId`

kidder commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,98 @@
+}  // namespace detail
+
+/*!
+ * \brief Registers itself with the local observer parallel component so the
```
`\brief Registers this element of a parallel component with the local `Observer` parallel component for each triggered observation.`

`\details This tells the `Observer` to expect data from this component, as well as whether each observation is a Reduction or Volume observation.`

kidder commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,98 @@
+ * \brief Registers itself with the local observer parallel component so the
+ * observer knows to expect data from this component, and also whether to expect
+ * Reduction, Volume, or both Reduction and Volume data observation calls.
+ */
```
add `Should be added to the phase dependent action list of the components that contribute data for volume and reduction observations`

kidder commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,98 @@
+      }
+    }
+
+    for (const auto& type_of_observation_and_observation_id_pair :
```
structured binding `[type_of_observation, observation_id]` ?

kidder commented on src/IO/Observer/Actions/RegisterSingleton.hpp:
````diff
@@ -0,0 +1,97 @@
+ * The singleton that observes is expected to call WriteReductionData
+ * on node 0. An example of how to do this is
+ *
+ *```
````
this needs to be converted to a snippet (or removed or also moved to the documentation of WriteReductionData; it has nothing to do with registration... )

kidder commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,98 @@
+ * observer knows to expect data from this component, and also whether to expect
+ * Reduction, Volume, or both Reduction and Volume data observation calls.
+ */
+struct RegisterEventsWithObservers {
```
RegisterEventsWithObservers -> RegisteredTriggeredObservationsWithObservers ?

kidder commented on src/IO/Observer/Actions/RegisterWithObservers.hpp:
```diff
@@ -0,0 +1,70 @@
+        *Parallel::get_parallel_component<observers::Observer<Metavariables>>(
+             cache)
+             .ckLocalBranch();
+    std::pair<observers::TypeOfObservation, observers::ObservationId>
```
structured binding?

kidder commented on src/IO/Observer/Actions/RegisterWithObservers.hpp:
```diff
@@ -0,0 +1,70 @@
+
+    Parallel::simple_action<observers::Actions::RegisterSenderWithSelf>(
+        observer,
+        std::move(                                                // NOLINT
```
remove the moves (to be consistent with the call in RegisterEvents.hpp)

kidder commented on src/IO/Observer/Actions/RegisterWithObservers.hpp:
```diff
@@ -0,0 +1,70 @@
+ * `std::pair<observers::TypeOfObservation, observers::ObservationId>`
+ */
+template <typename RegisterHelper>
+struct RegisterWithObservers {
```
RegisterObservationWithObservers?    or RegisterRequiredObservationsWithObservers?  Is this action going away or is  it how we are going to handle required observations (i.e. are all observations specified in the input file or are we allowing hard-coded one which is what this seems to handle)

nilsdeppe replied:
Ideally I want all observations to go through the events. The elliptic code does observations in some weird way that makes no sense to me, which is basically why this is staying. However, parts of the elliptic code are broken anyway, so hopefully once Nils F. fixes that we can remove this action. Or updating to use `std::vector` as you suggest below will be a lot easier.

kidder commented on src/IO/Observer/Actions/RegisterWithObservers.hpp:
```diff
@@ -0,0 +1,70 @@
+    std::pair<observers::TypeOfObservation, observers::ObservationId>
+        type_of_observation_and_observation_id_pair =
+            RegisterHelper::template register_info<ParallelComponent>(
+                box, array_index);
```
also it seems if this action is sticking around that it should specify a vector of pairs to make it more flexible...

nilsdeppe replied:
I agree but I don't feel like doing that... Someone else can do it

kidder commented on src/IO/Observer/Actions/RegisterSingleton.hpp:
````diff
@@ -0,0 +1,97 @@
+ *```
+ */
+template <typename RegisterHelper>
+struct RegisterSingletonWithObserverWriter {
````
can we just get rid of this action and handle the singleton case in `RegisterWithObservers` with a if constexpr?
Also we probably want an equivalent case in `RegisterEventsWithObservers` to handle triggered events on singletons?

nilsdeppe replied:
Maybe? I would like to remove this action in the long run, or have the functionality in another action that handles both singletons and arrays. Generalizing the events registration to work on singletons would be fairly easy, but I don't feel like doing it in this PR.

kidder commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -26,10 +26,10 @@
 #include "Evolution/Initialization/SetVariables.hpp"
 #include "Evolution/Systems/Burgers/System.hpp"
-#include "IO/Observer/Actions.hpp"
+#include "IO/Observer/Actions/ObserverRegistration.hpp"
```
why do I need this include?

kidder commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
-            contribute_count++;
           }
 
+          ASSERT(observations_registered.find(ObservationIdRegistrationKey{
```
why isn't this at the beginning of the function?

nilsdeppe replied:
Probably because when I hit the exception it wasn't until down here :shrug: I agree it should be at the start

kidder commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
+          } else if (UNLIKELY(current_names->second != reduction_names)) {
+            ERROR(
+                "The reduction names passed in must match the currently "
+                "known reduction names.");
```
print the mismatched names

nilsdeppe replied:
We should fix `StdHelpers` first so we don't suck in the entire STL in more places

kidder commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
+    }
+  }
+};
+
```
update the following documentation as (if I understand correctly) it is only used for writing on node 0

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -64,39 +65,69 @@ struct ContributeVolumeData {
       Requires<tmpl::list_contains_v<DbTagsList, Tags::TensorData>> = nullptr>
```
is the warning above still true?  If yes, what is the feature to be implemented in the future?

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -64,39 +65,69 @@ struct ContributeVolumeData {
-        ](const gsl::not_null<db::item_type<Tags::TensorData>*> volume_data,
-          const std::unordered_set<ArrayComponentId>&
-              volume_component_ids) mutable noexcept {
+        [&array_index, &cache, &observation_id, &sender_array_id,
```
can you please make (where possible) the argument names (for the apply function and the lambda in the mutate) consistent between this action and ContributeReductionData. It seems many of the ERROR/ASSERTs in the initial checks should be the same but the naming differences makes it hard to see if they are consistent.  You may be able to define a common helper function that can be called from both actions

nilsdeppe replied:
I will see about the names.

I don't like having a helper function for the ASSERT because then you don't know what actually triggered the ASSERT.

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -64,39 +65,69 @@ struct ContributeVolumeData {
+                  "The extents from the same volume component at a specific "
+                  "observation should always be the same. For example, the "
+                  "extents of a dG element should be the same for all calls to "
+                  "ContributeVolumeData that occur at the same time.");
```
print out the bad extents

nilsdeppe replied:
Again, need to fix stdhelpers first

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -129,59 +161,76 @@ struct ContributeVolumeDataToWriter {
+                volume_observers_contributed,
+            const std::unordered_map<ObservationIdRegistrationKey,
+                                     std::unordered_set<ArrayComponentId>>&
+                observations_registered) noexcept {
```
again please use consistent names (where possible) with the `CollectReductionDataOnNode` action

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -129,59 +161,76 @@ struct ContributeVolumeDataToWriter {
           // group. If so we write to disk.
-          if (volume_observers_contributed->at(observation_id) ==
-              expected_number_of_calls) {
+          if (volume_observers_contributed->at(observation_id).size() ==
```
why is this action not a ThreadedAction like `CollectReductionDataOnNode` with a bool, a locked block, and after the lock the call to the next action if the bool is true?  I find it helpful that whenever locks are used, there is an explanation of why they are needed

nilsdeppe replied:
I've changed this since @moxcodes wants oodles of locks flying around everywhere.

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp:
```diff
@@ -7,6 +7,7 @@
 
 #include "AlgorithmSingleton.hpp"
 #include "DataStructures/DataBox/DataBox.hpp"
+#include "IO/Observer/Actions/RegisterSingleton.hpp"
```
why the change to an include?  it is only used in a typelist

nilsdeppe replied:
Because things need to link

kidder commented on src/ParallelAlgorithms/Events/ObserveErrorNorms.hpp:
```diff
@@ -91,13 +95,20 @@ class ObserveErrorNorms<ObservationValueTag, tmpl::list<Tensors...>,
+  struct SubfileName {
+    using type = std::string;
+    static constexpr OptionString help = {
+        "The The name of the subfile inside the HDF5 file"};
```
remove a `The`

kidder commented on src/ParallelAlgorithms/Events/ObserveErrorNorms.hpp:
```diff
@@ -112,6 +123,7 @@ class ObserveErrorNorms<ObservationValueTag, tmpl::list<Tensors...>,
       "run at once will produce unpredictable results.";
```
is this warning (and the doxygen warning above) still true?

kidder commented on src/ParallelAlgorithms/Events/ObserveFields.hpp:
```diff
@@ -133,10 +141,14 @@ class ObserveFields<VolumeDim, ObservationValueTag, tmpl::list<Tensors...>,
       "triggered at a given time.  Causing multiple events to run at once\n"
```
again is this warning (and the above doxygen warning) still true?

kidder commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -106,6 +117,7 @@ class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
       "run at once will produce unpredictable results.";
```
same question about the warning...

kidder commented on tests/Unit/RunSingleTest/CMakeLists.txt:
```diff
@@ -32,14 +32,16 @@ add_spectre_executable(
   # The following lines list the source files with tests in them to be
   # compiled into the executable.
-  ../DataStructures/Test_DataVector.cpp
+  ../ParallelAlgorithms/LinearSolver/AsynchronousSolvers/Test_ElementActions.cpp
```
drop the changes to this file

kidder reviewed (changes requested):
I have yet to look carefully at the changes to the tests, but I browsed them.  Most of my requests are for improved documentation and names.  I realize you essentially moved some of the code; in such cases we can, if you wish, defer the changes to a subsequent PR if an issue is filed.   For discussion (but not a request for this PR) can we have a better name for the `ArrayComponentId`?  I find it confusing when it is used with non-array components.

nilsdeppe commented:
I will just improve all documentation now :) I'd rather get it done with while we're all thinking about it than going back.

I agree that `ArrayComponentId` is not a great not. It is used to identify an individual element of array-type components, and for a singleton there's just one element. `ComponentId` would be fine, but I've generally been unsure of what word(s) to use to distinguish between a component and an element of an array-type component. Ideally we'd be able to come up with a name for that. We could call the `members`. `Element` is already used and I worry people (including myself!) will get confused. Looking at a thesaurus, other terms would be `piece` and `constituent` that aren't completely opaque. Personally, I like either `member` (only potential confusion is member functions, but we use Actions and Entry Methods) and `constituent`.

kidder commented:
yes, agreed about distinguishing between an element of a component and a component; ComponentElementId I agree has the potential for confusion; I think I find ComponentMemberId acceptable; better than anything I've thought of so far

nilsdeppe commented:
@moxcodes @kidder I've pushed a few fixups mostly in an attempt to keep some of the changes a little factored out for easier reviewing. Thanks for the reviews! Please ping me on anything I missed. I tried to keep track, but I probably missed some or misinterpreted them.

moxcodes commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -19,251 +22,262 @@
-        });
+    if constexpr (tmpl::list_contains_v<
+                      DbTagList, observers::Tags::ObservationsRegistered>) {
+      // bool first_call_for_observation_id = false;
```
remove the commented code line

moxcodes commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,119 @@
+
+The observers infrastructure works with two parallel components: a group and a
+nodegroup. We have two types of observations
+`Reduction` and `Volume` (see the enum observers::TypeOfObservation). Reduction
```
backticks for `observers::TypeOfObservation`?

moxcodes commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,119 @@
+The observers infrastructure works with two parallel components: a group and a
+nodegroup. We have two types of observations
+`Reduction` and `Volume` (see the enum observers::TypeOfObservation). Reduction
+data is anything that is written
```
A few surprising newline choices

moxcodes commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,119 @@
+specified as the fourth template parameter to `Parallel::ReductionDatum`. See
+the documentation of `Parallel::ReductionDatum` for examples.
+
+At the start of a simulation, everything that wants to perform a reduction for
```
'everything' -> 'every component and event'?

moxcodes commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,119 @@
+purposes of observing data. The spectral coefficients can also be written as
+volume data, though some care must be taken in that case to correctly identify
+which mode is associated with which terms in the basis function
+expansion. Whatever component will contributed volume data to be written must
```
`will contributed`-> `will contribute`

moxcodes commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,119 @@
+`ContributeVolumeData` action. Once all the elements on a single core have
+contributed their volume data to the `observers::Observer` group, the
+`observers::Observer` group moves its data to the `observers::ObserverWriter`
+component to be written. We write one file per node append the node ID to the
```
The sentence 'We write one file per node append the node ID to the HDF5 file name' appears malformed, but I'm unsure what was intended.

moxcodes reviewed (commented):
I have a few details to suggest for the documentation and fixups.
I'm happy for you to squash the fixups and the additional changes I requested, as long is @kidder is similarly happy.

kidder commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,119 @@
+
+### Threading and NodeLocks
+
+Since the `observers::ObserverWriter` class is nodegroup, its entry methods can
```
is a nodegroup

kidder commented:
you can squash

nilsdeppe committed:
Rename RegisterSenderWithSelf

Renamed to RegisterContributorWithObserver

nilsdeppe committed:
Rename VolumesContributed

Renamed to ContributorsOfTensorData

nilsdeppe committed:
Rename ReductionObserversRegisteredNodes

Renamed to NodesExpectedToContributeReductions

nilsdeppe committed:
Rename NodeReductionsContributedForWriting

Rename to NodesThatContributedReductions

nilsdeppe committed:
Rename ObservationsRegistered

Renamed to ExpectedContributorsForObservations

nilsdeppe committed:
Rename ReductionsContributed

Renamed to ContributorsOfReductionData

nilsdeppe commented:
Squased, rebased on develop, and made a bunch of the suggested renames as additional comments. I think keeping renames in separate commits is easier, both in terms of squashing and reviewing :)

fmahebert commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,116 @@
+\cond NEVER Distributed under the MIT License.  See LICENSE.txt for details.
+\endcond
```
Does this work with the license text formatted in the usual way?
```
\cond NEVER
Distributed under the MIT License.
See LICENSE.txt for details.
\endcond
```

fmahebert commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,116 @@
+# Observers Infrastructure {#observers_infrastructure_dev_guide}
+
+The observers infrastructure works with two parallel components: a group and a
+nodegroup. We have two types of observations `Reduction` and `Volume` (see the
```
Add `:` after "observations"

fmahebert commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,116 @@
+
+The observers infrastructure works with two parallel components: a group and a
+nodegroup. We have two types of observations `Reduction` and `Volume` (see the
+enum observers::TypeOfObservation). `Reduction` data is anything that is written
```
Should the enum name be in backticks?

fmahebert commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,116 @@
+The observers infrastructure works with two parallel components: a group and a
+nodegroup. We have two types of observations `Reduction` and `Volume` (see the
+enum observers::TypeOfObservation). `Reduction` data is anything that is written
+once per simulation. Some examples of reduction data are integrals or L2 norms
```
This has the potential to be misleading: you've said reductions are for things written once, so perhaps it would be helpful to clarify that you mean an integral _at a particular time_? Otherwise this might be misunderstood to mean only one integral can be written during the simulation.

fmahebert commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,116 @@
+takes as template parameters the type of the data, a binary operator that
+combines two values into one, and optionally an n-ary operator that computes the
+final result using the extra reduction arguments specified as the fourth
+template parameter to `Parallel::ReductionDatum`. See the documentation of
```
I found this sentence difficult to read because it's quite abstract. I'm going to suggest moving some of its contents in the the paragraphs below... feel free to disagree with this suggestion.

-> "A `Parallel::ReductionDatum` takes as template parameters the type of the data, and operators that define how data from the different cores are to be combined to a single value. See the paragraphs below for more detail, and the documentation of `Parallel::ReductionDatum` for examples."

kidder commented:
@moxcodes please look at

fmahebert commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,278 @@
+/// `ObservationIdRegistrationKey` that will call
+/// `observers::ThreadedActions::ContributeVolumeData`.
+///
+/// Should be invoked on ObserverWriter by the the component that will be
```
typo: "by the the" -> "by the" ?

fmahebert commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,278 @@
+      const auto node_id = static_cast<size_t>(Parallel::my_node());
+      ASSERT(node_id == 0, "Only node zero, not node "
+                               << node_id
+                               << ", should be called from another node");
```
Can this message be made more specific? I worry that "should be called from" is a bit vague. Do you think it could be more useful to say something like "Only node zero, not node `node_id`, can be in charge of writing reduction data. But node `caller_node_id` tried to register with node `node _id` for writing" ?

fmahebert commented on src/IO/Observer/Actions/ObserverRegistration.hpp:
```diff
@@ -0,0 +1,278 @@
+                    const observers::ArrayComponentId& component_id,
+                    const TypeOfObservation& type_of_observation) noexcept {
+    if constexpr (tmpl::list_contains_v<
+                      DbTagList, observers::Tags::ObservationsRegistered>) {
```
Check clang-format? this formatting appears a bit different from that in the other actions... but maybe that's just clang-format doing its thing?

nilsdeppe replied:
Checked, it's because the things in the indent are inside the angle brackets. I'm not sure why it does 4 spaces instead of 2, but :shrug:

fmahebert commented on src/IO/Observer/Actions/RegisterEvents.hpp:
```diff
@@ -0,0 +1,135 @@
+        using EventType = typename decltype(event_type_v)::type;
+        if constexpr (detail::has_observation_registration_tags_v<EventType>) {
+          const auto* const derived_class_ptr =
+              dynamic_cast<const EventType*>(&event);
```
Can you add a short note of what is going on here with the casting? Is the idea that only one type EventType results in a valid (non-nullptr) result from the cast?

fmahebert commented on src/IO/Observer/ReductionActions.hpp:
```diff
@@ -76,90 +79,292 @@ struct ContributeReductionDataToWriter;
+      // reasonable efficient in parallel and so we manually guarantee that
+      // consistent state. To this end, we create pointers and assign to them
+      // the data in the DataBox which is guaranteed to be pointer stable. The
+      // data itself is guaranteed to be stable inside the ReductionDataLock.
```
Very useful comment 👍

fmahebert commented on src/IO/Observer/Tags.hpp:
```diff
@@ -58,6 +111,11 @@ struct ReductionDataNames;
 /// Reduction data to be written to disk.
+///
+/// If we have `M` `ArrayComponentIds` registered with a given `Observer`
+///  for a given `ObservationId`, then we expect the `Observer` to receive
```
[nitpick] there's an extra space at the start of this line

fmahebert commented on tests/Unit/IO/Observers/Test_RegisterSingleton.cpp:
```diff
@@ -0,0 +1,116 @@
+
+  CHECK(MockRegisterReductionContributorWithObserverWriter::result
+            .observation_key == observers::ObservationKey{"Singleton"});
+  // Need an `or` because we don't know what order actions are run in
```
Is this comment from a prior code version? I can't figure out what the "or" is that's being talked about.

fmahebert commented on docs/DevGuide/Observers.md:
```diff
@@ -0,0 +1,116 @@
+name/ID) has been contributed. Once all the data is collected on the core it is
+copied to the local `observers::ObserverWriter` nodegroup, which keeps track of
+how many of the cores on the node will be contributing to a specific
+observation. Once all the reductions on a node are complete, the final reduction
```
(continued from a prior comment)

-> "Once all the node's data is collected to the nodegroup, it is combined using the binary operator from `Parallel::ReductionDatum`'s second template parameter. Then, in a similar fashion, the final reduction is done to node `0`. ..."

... But I'm not sure if what I wrote above is correct, because I still have an unanswered question after reading your docs: at which stage(s) is the data combined using the binary op? per core? per node (this is what I think is going on from the code?)? once after everything is copied to node 0?

So regardless of whether you like my suggestion for redistributing the content of the abstract sentence above, I think this would be a good thing to clarify.

nilsdeppe replied:
I have expanded this discussion to clarify when things are combined. Please let me know what you think!

nilsvu commented on tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml:
```diff
@@ -46,11 +46,13 @@ EventsAndTriggers:
       Offset: 0
-  : - ObserveErrorNorms
+  : - ObserveErrorNorms:
+        SubfileName: Errors
```
Do you have a use case for changing the subfile name for observations in the input file? I wonder if it would be more user-friendly to have the subfile name be determined at compile-time, so it is the same for every run of an executable. Code that reads data from a particular executable's output can then rely on data being present at predictable locations in the subfile. What do you think?

nilsdeppe replied:
That doesn't work. If you want to be able to choose how often you want to observe and what to observe in the input file, you have to specify where you write it. Basically, this is the sole reason any of the changes actually allow you to have multiple reduction and volume observations running.

nilsvu replied:
Right of course 🤦‍♂️

nilsvu replied:
Wait I can't think of a use case now where you would have the same compile-time event observe multiple times, i.e. need to set the subfile name at runtime.. what would be an example?

nilsdeppe replied:
You want to dump the lapse every 10 time steps as volume data and the spacetime metric every 100. You want to dump the errors every 3 time steps and the constraints every 12. Literally anything where you want to have more than one event that does a reduction or observation, even if they are of different fields. This is what Tom tried to do: https://github.com/sxs-collaboration/spectre/issues/2363 

There are basically endless examples of where you'd want to observe different things at different frequencies, and requiring that be set at compile time would make the code horribly painful to use.

nilsvu replied:
What confuses me is that I thought the observations could be written into the same subfile even when they happen at different frequencies. If I understand you correctly, you would write the lapse into one subfile every 10 timesteps and the metric in another subfile every 100 timesteps. Can't they both be written in the same subfile, that always has the same name?

nilsvu replied:
Regarding #2363: To resolve that issue the subfiles don't need to be set at runtime. The `ObserveErrorNorms` writes into `error_norms.dat` and `ObserveVolumeIntegrals` into `volume_integrals.dat`, for example.

nilsdeppe replied:
No, they can't be because at every 100 time step you are getting data from 2 actions going into the same file. You now need to keep track of when you get data for only the lapse and when for both the lapse and the metric. This is going to make the HDF5 file format much more complicated than it currently is, and I don't really see any benefit to trying to do something that complicated. It's been the plan to use different subfile names since when we first discussed the observers like 4 years ago.

Sure, but is there what happens if you want to observe the error norms of the metric every 10 steps and the metric and auxiliary variables every 100 steps? This is not at all supported by the dat files and is basically impossible even at the HDF5 level without making the file format a complete nightmare. I guess you could write `0.0` at the times you don't observe it, but I expect people to be more confused that the error jobs to zero than that they are in different subfiles.

nilsvu replied:
Ok thanks for explaining 👍

nilsdeppe replied:
You're welcome! :)

fmahebert reviewed (commented):
I was pleasantly surprised that this is overall quite readable. I think the new dev guide you added is very helpful in giving an overview. I did have a few comments mostly relating to clarifying this documentation. Feel free to defer to a future PR if this review comes too late.

nilsvu commented on tests/Unit/ParallelAlgorithms/LinearSolver/ConjugateGradient/CMakeLists.txt:
```diff
@@ -20,5 +20,8 @@ add_dependencies(
+# Temporarily comment out until the elliptic solvers is updated to properly
+# register with observers. Currently an observation ID is used multiple times
+# by different events/components.
+# add_linear_solver_algorithm_test("ConjugateGradientAlgorithm")
```
Please apply this patch, then you can re-enable these tests:

```patch
diff --git a/src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ConjugateGradient.hpp b/src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ConjugateGradient.hpp
index 04dda3d84..9ebaaf096 100644
--- a/src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ConjugateGradient.hpp
+++ b/src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ConjugateGradient.hpp
@@ -112,8 +112,7 @@ struct ConjugateGradient {
    */
   using initialize_element = detail::InitializeElement<FieldsTag, OptionsGroup>;
 
-  using register_element =
-      observers::Actions::RegisterWithObservers<observe_detail::Registration>;
+  using register_element = tmpl::list<>;
 
   /*!
    * \brief Reset the linear solver to its initial state.
diff --git a/src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp b/src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp
index 56c985167..59bd5f08e 100644
--- a/src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp
+++ b/src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp
@@ -147,8 +147,7 @@ struct Gmres {
   using initialize_element =
       detail::InitializeElement<FieldsTag, OptionsGroup, Preconditioned>;
 
-  using register_element =
-      observers::Actions::RegisterWithObservers<observe_detail::Registration>;
+  using register_element = tmpl::list<>;
 
   /*!
    * \brief Reset the linear solver to its initial state.
```

I can deal with further refinements to how the parallel linear solvers do observations in follow-up PRs.

nilsdeppe commented:
I've pushed a fixup addressing the new comments. I included the patch to the elliptic code so all those tests are enabled again

kidder reviewed (commented):
you have my permission to squash

fmahebert reviewed (commented):
I'm also happy with the fixup.

moxcodes reviewed (commented):
feel free to squash

nilsdeppe commented:
I've rebased and squashed this.

I did need to comment out the two preconditioned elliptic tests (Gmres), but the other 4 tests seem to be working now.

fmahebert commented:
[Not a change request, but a suggestion for discussion:] Do you think it would be useful to update one (or more) of our input files to show this PR in action? One natural place to demonstrate more "features" of our observation might be in `PlaneWave1DObserveExample.yaml`, perhaps by dumping one of the other ScalarWave fields at a different frequency, or dumping different reductions at different frequencies. It might be nice to have an example to point users to. What do you think?

Edit to add: if we agree this would be a useful change, I'd be perfectly happy to see it done in a separate PR to accelerate merging this one

nilsdeppe commented:
Yea, I think that's reasonable. Here's what I changed in the `PlaneWave1DObserveExample.yaml`

```yaml
# [observe_event_trigger]
EventsAndTriggers:
  ? EveryNSlabs:
      N: 3
      Offset: 5
  : - ObserveErrorNorms:
        SubfileName: Errors
  ? SpecifiedSlabs:
      Slabs: [0, 100]
  : - ObserveFields:
        SubfileName: VolumePsi0And100
        VariablesToObserve: ["Psi"]
  ? EveryNSlabs:
      N: 50
      Offset: 0
  : - ObserveFields:
        SubfileName: VolumePsiPiPhiEvery50Slabs
        VariablesToObserve: ["Psi", "Pi", "Phi"]
  ? SpecifiedSlabs:
      Slabs: [100]
  : - Completion
# [observe_event_trigger]

```
I squashed right away and am happy to change or drop :)

fmahebert reviewed (dismissed):
Yeah this is exactly what I had in mind, thanks for the quick update!

nilsdeppe commented:
Great! Of course :) Having the features added here isn't useful if nobody knows about it :D

nilsvu commented on tests/Unit/ParallelAlgorithms/LinearSolver/Gmres/CMakeLists.txt:
```diff
@@ -20,7 +20,9 @@ add_dependencies(
+# updated to properly register with observers.
 add_linear_solver_algorithm_test("GmresAlgorithm")
-add_linear_solver_algorithm_test("GmresPreconditionedAlgorithm")
+# add_linear_solver_algorithm_test("GmresPreconditionedAlgorithm")
```
Here's the patch to fix these. I can put this in a follow-up PR if you prefer.

```diff
diff --git a/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/DistributedLinearSolverAlgorithmTestHelpers.hpp b/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/DistributedLinearSolverAlgorithmTestHelpers.hpp
index 7c5affe3b..7a444a555 100644
--- a/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/DistributedLinearSolverAlgorithmTestHelpers.hpp
+++ b/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/DistributedLinearSolverAlgorithmTestHelpers.hpp
@@ -328,6 +328,7 @@ struct ElementArray {
           typename Metavariables::Phase,
           Metavariables::Phase::RegisterWithObserver,
           tmpl::list<typename linear_solver::register_element,
+                     helpers::detail::register_preconditioner<preconditioner>,
                      typename linear_solver::prepare_solve,
                      Parallel::Actions::TerminatePhase>>,
       Parallel::PhaseActions<
diff --git a/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp b/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp
index b8008cf47..a508d632a 100644
--- a/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp
+++ b/tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp
@@ -222,6 +222,20 @@ template <typename Preconditioner>
 using init_preconditioner =
     typename init_preconditioner_impl<Preconditioner>::type;
 
+template <typename Preconditioner>
+struct register_preconditioner_impl {
+  using type = typename Preconditioner::register_element;
+};
+
+template <>
+struct register_preconditioner_impl<void> {
+  using type = tmpl::list<>;
+};
+
+template <typename Preconditioner>
+using register_preconditioner =
+    typename register_preconditioner_impl<Preconditioner>::type;
+
 template <typename Preconditioner>
 struct run_preconditioner_impl {
   using type =
@@ -271,6 +285,7 @@ struct ElementArray {
           typename Metavariables::Phase,
           Metavariables::Phase::RegisterWithObserver,
           tmpl::list<typename linear_solver::register_element,
+                     detail::register_preconditioner<preconditioner>,
                      typename linear_solver::prepare_solve,
                      Parallel::Actions::TerminatePhase>>,
       Parallel::PhaseActions<
```

nilsdeppe replied:
Okay, that seems to work!

nilsvu commented on src/IO/Observer/ObservationId.cpp:
```diff
@@ -7,8 +7,34 @@
+}
+
+std::ostream& operator<<(std::ostream& os, const ObservationKey& t) noexcept {
+  return os << '(' << t.key() << ')';
```
(not a change request, just something to consider improving) Since the `key` is the hash of the string, output is not very readable for the user. For example, this is the error message I got when debugging the linear solver observation:

```
Couldn't find registration key (11827073594539249116) in the registered observers.
```

If instead of the hash this printed the string, it would be more helpful. Of course this would mean storing the string instead of the hash, or both. What do you think?

nilsdeppe replied:
I agree. This is a small change so I'll make it.

nilsdeppe commented:
Okay, did the couple mentioned suggestions.
- Get the last couple elliptic tests working (thanks @nilsleiffischer )
- I changed `ObservationKey` to store both the hash and the string. The idea being this way we only hash once (and 8 bits for a `size_t` is a reasonably cheap amount of memory). I think having the string stored so it can be printed in the numerous diagnostics that are added in this PR is very useful (again, thanks @nilsleiffischer ). This only changed `ObservationId.hpp` and `ObservationId.cpp`. 
I squashed immediately since the changes are quite trivial and isolated.

nilsdeppe commented:
Okay, sorry for more noise on this... There were some clang-tidy complaints to sort out. Should be ready now
